File: leetcode/0942_di_string_match.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> diStringMatch(string S) {
    vector<int> result = {0};

    int max = 0, min = 0;
    for (const auto& c : S) {
      if (c == 'I') {
        result.push_back(++max);
      } else {
        result.push_back(--min);
      }
    }

    for (size_t i = 0; i < result.size(); i++) {
      result[i] -= min;
    }

    return result;
  }
};

int main() {
  auto solution = Solution();

  vector<int> expected = {2, 3, 1, 4, 0};
  vector<int> result = solution.diStringMatch("IDID");

  assert(expected == result);
}

File: leetcode/0942_di_string_match.py
# -*- coding: utf-8 -*-


class Solution(object):
    def diStringMatch(self, S):
        i, j, result = 0, 0, [0]

        for c in S:
            if c == 'I':
                i += 1
                result.append(i)
            else:
                j -= 1
                result.append(j)

        return [el - j for el in result]


if __name__ == '__main__':
    solution = Solution()

    assert [2, 3, 1, 4, 0] == solution.diStringMatch('IDID')
    assert [0, 1, 2, 3] == solution.diStringMatch('III')
    assert [2, 1, 0, 3] == solution.diStringMatch('DDI')
